### DISCUSSION TITLE
ModuleFederationPlugin can the "remotes" prop be dynamic?

### DISCUSSION THREAD

**Discussion by jazzfog** [webpack/webpack#12213] [Q&A]
Hi
With the `remotes` property we define where to pull remotes from:
```js
remotes: {
  widget: `widget@http://localhost:3001/remoteEntry.js`,
},
```

It is possible to alternate the URL during the build
```js
remotes: {
  widget: `widget@${isDev ? 'http://localhost:3001' : '/apps/widgets'}/remoteEntry.js`,
},
```
Though I do not see so far a way to define the remote's URL dynamically, when application is actually running. 
I would imagine it like this:
```js
remotes: {
  widget: () => {
    const prefix = location.host.startsWith('eu.') ? 'europe' : 'us';
    return `widget@${prefix}/apps/widgets/remoteEntry.js`
  },
},
```
Is there a way to compute the target URLs?
Thanks!

**Comment by alexander-akait** [comment]
I think better is move this logic in runtime

**Reply by jazzfog** [reply]
Yes, this is what I am trying to achieve. Sorry if I did not make it clear.
I'd like to be able to compute the paths in runtime. 
How can I do that?

**Reply by alexander-akait** [reply]
You can use async `import`, depend on your runtime logic (`locale`/`window.location`/`domain`/etc), hard to say without small example

**Reply by jazzfog** [reply]
The logic/condition I will apply is irrelevant.
I am interested in finding a way to tell the 'ModuleFederationPlugin' (in runtime) where to request the federated modules from.

**Reply by alexander-akait** [reply]
Why do not create multiple `remotes` - `widget-eu` and `widget-us` and use it:
```js
// Pseudo code
async function loadWidget(condition) {
  return condition() ? import('widget-eu') : import('widget-us');
}

loadWidget(() => window.location.host.endsWith('eu'));
```

**Reply by jazzfog** [reply]
Oh, I probably oversimplified the example.
In reality the number of hosts is dynamic. It can be computed/pulled from somewhere in runtime but cannot be hardcoded beforehand in the config...

**Reply by alexander-akait** [reply]
It's not possible to collect them at the build stage, right?

**Reply by jazzfog** [reply]
Unfortunately not...

**Reply by alexander-akait** [reply]
technically I think it's possible

**Reply by jazzfog** [reply]
Sadly not always. 
The eu/us example I gave is just a simplification, in the reality it is more complex.

**Reply by alexander-akait** [reply]
I mean technically it's possible to implement in webpack

**Reply by jazzfog** [reply]
Uh, sorry, yes this is what I thought, at least from what I understand there should be nothing to prevent it I just can't seem to find a way how to go about it.

**Reply by alexander-akait** [reply]
We need implement this

**Reply by jazzfog** [reply]
Created the issue: https://github.com/webpack/webpack/issues/12217
